pkg/skaffold/deploy/helm/dependencygraph.go

```diff
@@ -0,0 +1,198 @@
+	if !g.hasDependencies {
+		return nil
+	}
+
+	visited := make(map[string]bool)
+	recStack := make(map[string]bool)
+
+	var checkCycle func(node string) error
+	checkCycle = func(node string) error {
+		if !visited[node] {
+			visited[node] = true
+			recStack[node] = true
+
+			for _, dep := range g.graph[node] {
+				if !visited[dep] {
+					if err := checkCycle(dep); err != nil {
+						return err
+					}
+				} else if recStack[dep] {
+					return fmt.Errorf("cycle detected involving release %s", node)
+				}
+			}
+		}
+		recStack[node] = false
+		return nil
+	}
+
+	for node := range g.graph {
+		if !visited[node] {
+			if err := checkCycle(node); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// GetReleasesByLevel returns releases grouped by their dependency level while preserving
+// the original order within each level. Level 0 contains releases with no dependencies,
+// level 1 contains releases that depend only on level 0 releases, and so on.
+func (g *DependencyGraph) GetReleasesByLevel() (map[int][]string, error) {
+	if len(g.releases) == 0 {
+		// For empty releases, return empty map to avoid nil
+		return map[int][]string{}, nil
+	}
+
+	if !g.hasDependencies {
+		// Fast path:  if no dependencies, all releases are at level 0
+		// Preserve original order from releases slice
+		return map[int][]string{
+			0: g.getNames(),
+		}, nil
+	}
+
+	// Check for cycles before calculating deployment order
+	if err := g.HasCycles(); err != nil {
+		return nil, err
+	}
+
+	order, err := g.calculateDeploymentOrder()
+	if err != nil {
+		return nil, err
+	}
+
+	return g.groupReleasesByLevel(order), nil
+}
+
+// getNames returns a slice of release names in their original order
+func (g *DependencyGraph) getNames() []string {
+	names := make([]string, len(g.releases))
+	for i, release := range g.releases {
+		names[i] = release.Name
+	}
+	return names
+}
+
+// calculateDeploymentOrder returns a topologically sorted list of releases,
+// ensuring that releases are deployed after their dependencies while maintaining
+// the original order where possible
+func (g *DependencyGraph) calculateDeploymentOrder() ([]string, error) {
+	visited := make(map[string]bool)
+	order := make([]string, 0)
+
+	var visit func(node string) error
+	visit = func(node string) error {
+		if visited[node] {
+			return nil
+		}
+		visited[node] = true
+
+		for _, dep := range g.graph[node] {
+			if err := visit(dep); err != nil {
+				return err
+			}
+		}
+		order = append(order, node)
+		return nil
+	}
+
+	// Process releases in their original order
+	for _, release := range g.releases {
+		if err := visit(release.Name); err != nil {
+			return nil, err
+		}
+	}
+
+	return order, nil
+}
+
+// groupReleasesByLevel groups releases by their dependency level while preserving
+// the original order within each level
+func (g *DependencyGraph) groupReleasesByLevel(order []string) map[int][]string {
+	levels := make(map[int][]string)
```

Comment:
@alphanota and, here we have a map, but the value is slice, and the slice will have the  original order

Reply:
> All the releases in the same level do not depend on each other, so the order in which they are installed should not matter

Why do you think so? They could depend on each other implicitly, and in my case, they depend on each other implicitly - first we have to install the release that comes first in the config order, then second, then third and so on